content/blog/2017/2017-02-10-declarative-html-publisher.adoc

```diff
@@ -0,0 +1,322 @@
+
+At the Pipeline level,
+Declarative instead supports a limited set of Pipeline-specific controls:
+link:/doc/book/pipeline/syntax/#when[`when` directive], which we'll look at in
+a later blog in this series, control whether to execute the steps in a stage,
+and
+link:/doc/book/pipeline/syntax/#post[`post` sections]
+control which actions to take based on result of a single stage
+or a whole Pipeline.  `post` supports a number of
+link:/doc/book/pipeline/syntax/#post-conditions[run conditions],
+including `always` (execute no matter what) and `changed`
+(execute when the result differs from previous run).
+We'll use `success` to run `archive` and `publishHTML` when their respective stages complete.
+We'll also use an `always` block with a placeholder for sending notifications,
+which I'll implement in the next blog post.
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent any
+  options {
+    // Only keep the 10 most recent builds
+    buildDiscarder(logRotator(numToKeepStr:'10'))
+  }
+  stages {
+    stage ('Install') {
+      steps {
+        // install required gems
+        sh 'bundle install'
+      }
+    }
+    stage ('Build') {
+      steps {
+        // build
+        sh 'bundle exec rake build'
+      }
+
+      post {
+        success {
+          // Archive the built artifacts
+          archive includes: 'pkg/*.gem'
+        }
+      }
+    }
+    stage ('Test') {
+      steps {
+        // run tests with coverage
+        sh 'bundle exec rake spec'
+      }
+
+      post {
+        success {
+          // publish html
+          publishHTML target: [
+              allowMissing: false,
+              alwaysLinkToLastBuild: false,
+              keepAll: true,
+              reportDir: 'coverage',
+              reportFiles: 'index.html',
+              reportName: "RCov Report"
+            ]
+        }
+      }
+    }
+  }
+  post {
+    always {
+      echo "Send notifications for result: ${currentBuild.result}"
+    }
+  }
+}
+// Scripted //
+----
+
+== Switching `agent` to run in Docker
+
+`agent` can actually accept
+link:link:/doc/book/pipeline/syntax/#agent-parameters[several other parameters] instead of `any`.
+We could filter on `label "some-label"`, for example,
+which would be the equivalent of `node ("some-label")` in Scripted Pipeline.
+However, `agent` also lets us just as easily switch to using a Docker container,
+which replaces a more complicated set of changes in Scripted Pipeline:
+
+////
+Since this is a partial, non-copyable, Declarative Pipeline it shouldn't use
+[pipeline] so it's not presented as a Jenkinsfile
+////
+[source,groovy]
+----
+pipeline {
+  agent {
+    // Use docker container
+    docker {
+      image 'ruby:2.3'
+    }
+  }
+  /* ... unchanged ... */
+}
+----
+
+If I needed to I could add a `label` filter under `docker`
+to select a node to host the Docker container.
+I already have Docker host available on my agents, so I don't need it - this works as is.
```

Comment:
*I already have Docker host available on my agents* Is that really how to phrase this?